internals-js/src/specs/sourceSpec.ts

```diff
@@ -140,8 +144,271 @@ export class SourceSpecDefinition extends FeatureDefinition {
   sourceFieldDirective(schema: Schema) {
     return this.directive<SourceFieldDirectiveArgs>(schema, 'sourceField')!;
   }
+
+  override validateSubgraphSchema(schema: Schema): GraphQLError[] {
+    const sourceAPIInSchema = schema.directive('sourceAPI') as DirectiveDefinition<SourceAPIDirectiveArgs> | undefined;
+    const sourceTypeInSchema = schema.directive('sourceType') as DirectiveDefinition<SourceTypeDirectiveArgs> | undefined;
+    const sourceFieldInSchema = schema.directive('sourceField') as DirectiveDefinition<SourceFieldDirectiveArgs> | undefined;
+
+    if (!(sourceAPIInSchema || sourceTypeInSchema || sourceFieldInSchema)) {
+      // If none of the @source* directives are present, nothing needs
+      // validating.
+      return [];
+    }
+
+    const errors: GraphQLError[] = [];
+    const apiNameToProtocol = new Map<string, ProtocolName>();
+
+    if (sourceAPIInSchema) {
+      sourceAPIInSchema.applications().forEach(application => {
+        const { name, ...rest } = application.arguments();
+
+        if (apiNameToProtocol.has(name)) {
+          errors.push(new GraphQLError(`${sourceAPIInSchema.name} must specify unique name`));
+        }
+
+        // Ensure name is a valid GraphQL identifier.
+        if (!/^[a-zA-Z_][0-9a-zA-Z_]*$/.test(name)) {
+          errors.push(new GraphQLError(`${sourceAPIInSchema}(name: ${
+            JSON.stringify(name)
+          }) must be valid GraphQL identifier`));
+        }
+
+        let protocol: ProtocolName | undefined;
+        KNOWN_SOURCE_PROTOCOLS.forEach(knownProtocol => {
+          if (rest[knownProtocol]) {
+            if (protocol) {
+              errors.push(new GraphQLError(
+                `${sourceAPIInSchema.name} must specify only one of ${KNOWN_SOURCE_PROTOCOLS.join(', ')}`,
+              ));
+            }
+            protocol = knownProtocol;
+          }
+        });
+
+        if (protocol) {
+          apiNameToProtocol.set(name, protocol);
+
+          const protocolValue = rest[protocol];
+          if (protocolValue && protocol === "http") {
```

Comment:
Make 'http' a const that exists in KNOWN_PROTOCOLS?

Reply:
https://github.com/apollographql/federation/pull/2910/commits/633e78787b2b98b03647e23d2a2c1744aecdbb2e